[CMake] Simplify add_swift_library

llvm-beanz opened the pull request:
This patch splits add_swift_library into two functions one which handles the simple case of adding a library that is part of the compiler being built and the second handling the more complicated case of "target" libraries, which may need to build for one or more targets.

Most of the files changed in this patch are just to have the named parameters of the new add_swift_library function match the named parameters of llvm_add_library so that the LLVM CMake function can be re-used, which further simplifies the implementation of add_swift_library.

Commit:
[CMake] Simplify add_swift_library

This patch splits add_swift_library into two functions one which handles
the simple case of adding a library that is part of the compiler being
built and the second handling the more complicated case of "target"
libraries, which may need to build for one or more targets.

The new add_swift_library is built using llvm_add_library, which re-uses
LLVM's CMake modules. In adapting to use LLVM's modules some of
add_swift_library's named parameters have been removed and
LINK_LIBRARIES has changed to LINK_LIBS, and LLVM_LINK_COMPONENTS
changed to LINK_COMPONENTS.

This patch also cleans up libswiftBasic's handling of UUID library and
headers, and how it interfaces with gyb sources.

add_swift_library also no longer has the FILE_DEPENDS parameter, which
doesn't matter because llvm_add_library's DEPENDS parameter has the same
behavior.

llvm-beanz commented:
@swift-ci please test

erg commented:
LGTM!

gottesmm commented:
@erg I asked Chris to add more details to his commit msg. But yes, then it LGTM to me as well.

gottesmm reviewed (changes requested):
Please extend the commit message in the manner in which we spoke in person.

gottesmm commented:
Looks beautiful to me!

gottesmm commented:
@swift-ci Please clean test and merge

gottesmm commented:
@swift-ci Please clean test and merge

hughbe commented on lib/AST/CMakeLists.txt:
```diff
@@ -55,11 +55,10 @@ add_swift_library(swiftAST STATIC
     swiftMarkup
     swiftBasic
 
-  INTERFACE_LINK_LIBRARIES
```
This is necessary to fix a Windows linker error - clangDriver links to version.lib. If we used LINK_LIBRARIES, as we did before, we wouldn't transitively get the version.lib symbols. 
with INTERFACE_LINK_LIBRARIES, we do

Do we still maintain this behaviour?

llvm-beanz replied:
`LINK_LIBRARIES` as implemented by Swift's `_add_swift_library_single` required a separate dentation for interface dependencies. LLVM's `llvm_add_library` does not because it does the right thing for `LINK_LIBS`. Any library dependencies specified on static archives become interface dependencies.

I would expect this patch to work correctly on Windows, for the same reason LLVM has never had the problem that you were working around.

hughbe replied:
Okay, thanks

erg commented:
```
15:45:46 
lib/liblldbBase.a(lldb.cpp.o):/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/lldb/source/lldb.cpp:function lldb_private::GetVersion(): error: undefined reference to 'swift::version::getSwiftNumericVersion()'
```

Is this unrelated? If so I would merge.

erg commented:
@swift-ci Please clean test Linux platform and merge

gottesmm commented:
@swift-ci Please clean test Linux platform

gottesmm commented:
@erg The *and* merge functionality doesn't work with single platform tests.

llvm-beanz commented:
@erg, I don't believe that build issue is related, but I'm investigating now, so I'd rather not merge this until I have figured that out.

llvm-beanz commented:
I've managed to build this PR on FreeBSD using build-script and ld.gold, so I'm pretty confident that the Linux failure was unrelated. Just to be sure, let's try it again.

@swift-ci please clean test linux platform

llvm-beanz commented:
I figured out the issue causing linux failures.

@swift-ci please clean test and merge.

llvm-beanz commented:
@swift-ci please clean test

gottesmm commented:
@llvm-beanz Can you rebase and add a more descriptive message on what the issue was in the Linux case?

llvm-beanz commented:
@swift-ci please clean test

llvm-beanz commented:
@swift-ci please clean test OS X platform

Commit:
[CMake] Fix up an issue encountered in Linux PR testing

This should fix out-of-tree LLDB builds.

This patch overrides the output directories setup by llvm_add_library.
This is required because swift deviates from the LLVM convention of using the LLVM_*_INTDIR variables.

This was causing failures on LLDB's PR testing because LLDB has a terrible mechanism for picking the swift libraries. It finds all the .a files by searching a directory.

llvm-beanz commented:
Ok... I think I have force-pushed the right thing now... Gah!

@swift-ci please clean test

gottesmm commented:
@llvm-beanz LGTM! Thanks for fixing the commit messages!
